Sketcher: Change enter behavior on OVP to put OVP in lock state only

tetektoza opened the pull request:
This patch adds/changes a couple of things:
* if you press enter on a label now, it moves you to another label and adds the label and lock on the previous label, instead of previous behavior where it was accepting whole dimension
* if you press enter and have lock state on both labels then you move to next stage
* if you press ctrl+enter it's as is if you'd press enter on both labels (the object becomes constrained with whatever dimensions that were in both labels)
* tab still works the same way
* you can remove "Lock" state from the label by typing something additional or removing the dimension at all
* and of course mouse works independently from this update 

Demo:

https://github.com/user-attachments/assets/012b5365-fb83-4731-a5d5-2c03da8b43d7

Resolves: https://github.com/FreeCAD/FreeCAD/issues/21927

PaddleStroke commented on src/Mod/Sketcher/Gui/DrawSketchHandlerArc.h:
```diff
@@ -729,7 +729,7 @@ void DSHArcController::doChangeDrawSketchHandlerMode()
             auto& fourthParam = onViewParameters[OnViewParameter::Fourth];
 
-            if (thirdParam->hasFinishedEditing || fourthParam->hasFinishedEditing) {
+            if (thirdParam->hasFinishedEditing && fourthParam->hasFinishedEditing) {
```
This is what caused the bug. I checked the history and indeed it was && before. Which was introduced by https://github.com/FreeCAD/FreeCAD/pull/20925
I have not looked at 20925 in details. Why did you made this change initially? Was it a necessity? Just trying to understand why this present PR is so large.

tetektoza replied:
We agreed upon this change. Since we didn't have any indicator if field is accepted (`hasFinishedEditing` state) or not, we decided to make Enter accept all fields and go further. It could also be done and possible to do so enter would accept first field and then accept next field too and constrain with whatever dimension was in the next label (instead of just accepting the geometry and not constraining second label), but we didn't discuss it further. So now I've reverted the change and added a padlock icon to indicate to user which field actually has the locked state.

tetektoza commented:
@PaddleStroke If you press something on the keyboard, the dimension is becoming *ONLY* constrained and this was done to lock input to prevent from changing values in the spinbox by mouse movement (this was the easiest approach and using already made mechanisms instead of inventing new ones and potentially breaking OVP). So, for typing we set `isSet` state. For actually accepting the dimension, we use `hasFinishedEditing` state. If all labels on current stage have `hasFinishedEditing` state, then we go to another stage.

But since the dimension is becoming only constrained if you type but it still goes through all the necessary logic (and the same logic as when user presses enter), we had to know when user wants to finish action on the label, that's why the second state. Now, if you press enter - it accepts whatever is on the label and goes to another one (or next stage if there is only one label), but now since I've introduced TAB key to allow to switch between labels, user can go back to previous label and edit. So since they can switch multiple times between labels, user does not know which label has which state. Instead of forcing the user to go through each label again and press enter on each of them or remember the state which label had which state, I've introduced a padlock icon (as discussed with other guys).

So basically, typing doesn't do anything. It just constraints the dimension, the behavior is different from previous behavior. Now to actually accept it you have to press enter, but there was no visual distincion *when user pressed enter on which label*, so I've changed this `&&` statement to `||` statement to accept everything with enter. You didn't like it, so I've added a padlock icon to indicate that the label has this state upon pressing enter, and to enter everything faster I've added a shortcut of CTRL+ENTER to accept everything in a haul in current stage. An alternative to this approach (to also satisfy your github issue that you've put) is to remove padlock icon and make enter accept whatever is in any other label and constrain the dimension with whatever value was in the label.

But personally, I don't think this padlock icon is invasive, I've added alternatives for user to completely bypass this icon, it's just a small indication, and I think other CAD-related programs have similar approaches (like Fusion with a padlock icon).

PaddleStroke commented:
> If you press something on the keyboard, the dimension is becoming ONLY constrained

What does that mean?

PaddleStroke commented:
The fact that I had to read 3 times to understand is imo an indication that there is something fishy going on.

So if I understand correctly : 
- Red color is still an indication of isSet. But not hasFinishedEditing.
- An OVP can be 'isSet' but not 'hasFinishedEditing'. This can happen if the user start typing, let's say '10' then he hits tab, that goes to the 2nd OVP. There he types something press Enter.

In this case we have : 
- OVP1 : isSet = true  hasFinishedEditing = false 
- OVP2 : isSet = true  hasFinishedEditing = true

Then because the user pressed enter, hasFinishedEditing is false so it does not moves to the next mode. Is that correct?

So to fix this you want to introduce a lock icon, so when this case happen and enter doesn't move to next mode, the user has to see there is a missing lock icon and has to go to that OVP and press enter?

This is not a good solution:
- The OVP have now multiple status. The user doesn't understand why OVP becomes red and it doesn't work. He has to see there's a missing lock icon...
- It adds additional complexity.

On top of that, before people had specifically asked me that tab validate the OVP same as Enter. So why don't you simply put hasFinishedEditing to true when user press tab? Then no lock icon needed.

PaddleStroke commented:
Pressing tab should set hasFinishedEditing only is isSet = false. 

> But people asked specifically for tab to allow to cycle between labels and to be able to erase those states in previous labels.

You have to ponder the importance of a feature request to the consequences it has before implementing it.
Here being able to cycle back to erase a previously entered OVP is particularly not worth introducing https://github.com/FreeCAD/FreeCAD/issues/21927 nor the complexity of a dual state of the OVP that is going to add a lot of confusion.

If you don't think I am right, the simple answer is usually to look at other cad softwares (SW f360) and how they do it. I don't have them to test but I'd be surprised if they have this kind of validated-but-not-really kind of state.

PaddleStroke commented:
Also I have been thinking about the original issue #13089.

And was thinking that another solution to that initial issue would perhaps be less problematic : 
- Add bool 'beingEdited'
- On first key stroke, beingEdited = true. (isSet stays false).
- On mouse move, if beingEdited, then we do not update the value of the OVP -> data is not overriden fixing 13089.
- When editing is done -> isSet = true and beingEdited = false

This way we solve 13089 without having to detect backspace to cancel the isSet. Then those other issues are also gone.

hyarion commented:
Most of paddle's feedback has been addressed, dwg are happy with the changes and code good too, merging this.

PaddleStroke commented:
> Most of paddle's feedback has been addressed, dwg are happy with the changes and code good too, merging this.

There is no 3 states for the OVP anymore? Well I'll rebase and see for myself soon enough I guess. But I hope this is not false advertizing...
